### ISSUE TITLE
results inconsistency

### ISSUE THREAD

**Issue by NoaGarnett** [lorenmt/mtan#56]
Hi,
Many thanks for the paper and for the most readable code!
I tried to reproduce two lines out of table 3 in the paper using your code and following the instructions in the readme file.
The results I got are not consistent with those in the paper (not sure how to attach - attaching as spreadsheet and as image):
![Table_3](https://user-images.githubusercontent.com/10104853/148905209-e380d6f0-fbd0-4fda-a8a8-168fda675327.png)
[mtan_res.ods](https://github.com/lorenmt/mtan/files/7844776/mtan_res.ods)

All number are average of validation results for the last 10 epochs, as printed by the code.
How can such discrepancy be explained?

**Comment by lorenmt** [comment]
Hello,
 
Does this inconsistency occur in other networks as well?

At the moment, I would suggest following stronger baseline implementations which apply data augmentation on NYUv2 + ResNet-50 backbone + benchmarking using average relative performance -- which are now the standard in MTL research.

The results in my original paper is a bit outdated.

**Comment by lorenmt** [comment]
Though I agree that the performance of MTAN varies based on the task/data and other optimisation techniques. But from my experience and the feedback from other papers, MTAN should produce consistent improvement over Split in a non-trivial margin. 

The average relative performance is simply a metric to measure multitask performance in a singular value. But yes, I would first build MTAN on top of ResNet50 as the start, since it's a much stronger backbone after all.

**Comment by lorenmt** [comment]
Besides, honestly speaking, this is the very first time I heard anyone claiming that MTAN is worse than Split baseline. Though the reported numbers you attached look reasonable, this is definitely not common generally.

**Comment by NoaGarnett** [comment]
Thanks you again. This is really strange - I used your code as is. Anyway, I'll check with Resnet50 and update.
Regards,
Noa Garnett

**Comment by NoaGarnett** [comment]
Hi again,
I'm training NYUv2 using resnet50 with / without MTAN, and have a question regarding the decoder implementation in MTANDeepLabv3, and specifically about ASPPPooling:
The module first perform global average pooling, which makes each input a single vector, then Conv2d which is actually a linear ("fully-connected") layer, since no spatial information exists, then BatchNorm2d.
The input to BatchNorm2d has no spatial dimension, so the number of inputs to be normalized, for each channel, is equal to the batch size. If borrowing hyper parameters from Segnet training script, where batch size is 2, we are left with normalization of two inputs. Does it make sense?
I enlarged batch size to 4 on MTANDeepLabv3 training (kept other parameter as before - learning rate, scheduling, number of epochs).

**Comment by lorenmt** [comment]
I am not exactly sure what's your concern here. The batchnorm should work similar to the SegNet, the ASPP module simply aggregates information from different spatial dimension with different conv layers built with different dilation rates. And yes you can change batch size to any number you want.

**Comment by NoaGarnett** [comment]
Results using Resnet50 (did not change evaluation method) attached.
![Table_3_3](https://user-images.githubusercontent.com/10104853/149275329-c9ba5030-6838-40a6-bc7e-70354abc7bed.png)
[mtan_res.ods](https://github.com/lorenmt/mtan/files/7860355/mtan_res.ods)

Regards,
Noa

**Comment by lorenmt** [comment]
Did you apply pre-trained ImageNet features? The parameters was designed for non-pre-trained features.

**Comment by NoaGarnett** [comment]
I initialized with torchvision resnet50 weights, as in the code.

**Comment by lorenmt** [comment]
Yes, sorry. You should have turned that off.

**Comment by NoaGarnett** [comment]
OK, I'll turn it off and retry. Going on vacation - will update in 10 days.
Many thanks.

**Comment by lorenmt** [comment]
By the way, how you designed the Split, Wide for ResNet-50?

**Comment by lorenmt** [comment]
To follow up on this:

As you claim is quite serious, I have rerun the experiments and I cannot reproduce your results.

My results:
For Split Wide [Last 10 Epochs]; Training Script: `python3 model_segnet_split.py --gpu 0 --apply_augmentation --type wide`
<img width="909" alt="image" src="https://user-images.githubusercontent.com/6267679/149619931-11e769e2-3e4d-4ba5-a023-fd783e5eb64a.png">

For MTAN [Last 10 Epochs]; Training Script: `python3 model_segnet_mtan.py --gpu 0 --apply_augmentation`
<img width="914" alt="image" src="https://user-images.githubusercontent.com/6267679/149619913-2d912dc2-608d-45b6-ba28-14cd38a9b8a7.png">

Both networks are built on top of SegNet, with augmentation, and with no modification on the hyper-parameters. MTAN clearly outperforms Split, Wide by a large margin in all 3 tasks.

**Comment by NoaGarnett** [comment]
Indeed I was wrong - ResNet-50 experiment should not have been labeled Split, Wide - I mistakenly copied the label - should be Split, vanilla - I did not change the number of parameters in the backbone.

**Comment by NoaGarnett** [comment]
Hi again,
My friend got results similar to mine.
Basically, we get results similar to yours when using MTAN, and better than yours with Split, Wide configuration.
Can you send your python packages configuration, that is output of 
_**pip freeze**_ if using virtual environment, or _**conda env export**_ if using conda? The difference must hide somewhere there.
I attach the configuration of our environment.
[requirements.txt](https://github.com/lorenmt/mtan/files/7932751/requirements.txt)

Thanks and regards

**Comment by lorenmt** [comment]
Any new updates?

**Comment by NoaGarnett** [comment]
Yes. My friend got results similar to mine. We are preparing a [google colab](https://colab.research.google.com/) notebook showing the full process we did, starting from git clone, data download and training. We will also share the versions of relevant python packages. Once we're done preparing this notebook (training currently running) we'll share it here, hoping to find the source of results discrepancy. Basically, we get similar results to yours running MTAN, and significantly better then yours (similar to MTAN results, except on surface normals task) using Split, Wide config.

**Comment by NoaGarnett** [comment]
Hi again.
We created two colab notebook, each showing the entire process, starting from git clone, with train results.
[here](https://colab.research.google.com/drive/17691d4PyUjqDmwdjBC-J_UmZcxR8qeuf?usp=sharing#scrollTo=vjGyPWnT-r5g) is the first, starting with MTAN train (completed), and then <Split,Wide> config - currently still running, will be done in about two hours - you can refresh and check.
[here](https://colab.research.google.com/drive/1A4o4ZpVoSsoZ8Nh7XBICMd2FGe9fh20t?usp=sharing#scrollTo=QRvRERS2fu-7) is the second - starting with <Split, Wide> (completed), then running MTAN (still running, almost done).

Both configs has some variance is their results, and in general MTAN results is not better than <Split,Wide>

If you want write permission to these notebooks, so that you can play with it yourself, please let me know.

Please let us know which version of the relevant packages you are using - I can think of no other source for discrepancy.

Regards,
Noa

**Comment by lorenmt** [comment]
Thanks for spending time on this. My training results have also attached for [MTAN](https://www.dropbox.com/s/yjgluea41ree9yq/mtan.txt?dl=0) and [Split Wide](https://www.dropbox.com/s/04h7owb4pjx609u/split_wide.txt?dl=0).

But I have reproduced your results on another machine.

It seems like the performance for Split Wide is not stable.  

I wouldn't suggest spending additional time on this... as the performance will also depend on dataset and other hyper-parameters.

Let me know if you have further questions.
